### PR TITLE
[DebugInfo][IndVarSimplify] Fix missing debug location updates

### LINKED ISSUE #91436: [DebugInfo][IndVarSimplify] Missing debug location updates of newly created instructions
I've found missing debug location updates in IndVarSimplify Pass (IndVarSimplify[-L358](https://github.com/llvm/llvm-project/blob/6f2997cefc1e32c11a891ede2e3a2d73310e6ce1/llvm/lib/Transforms/Scalar/IndVarSimplify.cpp#L358), [-L363](https://github.com/llvm/llvm-project/blob/6f2997cefc1e32c11a891ede2e3a2d73310e6ce1/llvm/lib/Transforms/Scalar/IndVarSimplify.cpp#L363), [-L368](https://github.com/llvm/llvm-project/blob/6f2997cefc1e32c11a891ede2e3a2d73310e6ce1/llvm/lib/Transforms/Scalar/IndVarSimplify.cpp#L368C3-L370C64), [-L394](https://github.com/llvm/llvm-project/blob/6f2997cefc1e32c11a891ede2e3a2d73310e6ce1/llvm/lib/Transforms/Scalar/IndVarSimplify.cpp#L394)), and used -debugify on the regression test `floating-point-iv.ll` to double check the missing debug locations:

```
WARNING: Instruction with empty DebugLoc in function test1 --  %indvar.conv = sitofp i32 %iv.int to double
WARNING: Instruction with empty DebugLoc in function test1 --  %.int = add nuw nsw i32 %iv.int, 1
WARNING: Instruction with empty DebugLoc in function test1 --  %1 = icmp ult i32 %.int, 10000

WARNING: Instruction with empty DebugLoc in function test2 --  %indvar.conv = sitofp i32 %iv.int to double
WARNING: Instruction with empty DebugLoc in function test2 --  %.int = add nsw i32 %iv.int, 2
WARNING: Instruction with empty DebugLoc in function test2 --  %1 = icmp slt i32 %.int, -1

WARNING: Instruction with empty DebugLoc in function test5 --  %.int1 = add nsw i32 %.int, -1
WARNING: Instruction with empty DebugLoc in function test5 --  %3 = icmp slt i32 %.int1, 0

WARNING: Instruction with empty DebugLoc in function test_max_be --  %indvar.conv = sitofp i32 %tmp11.int to double
WARNING: Instruction with empty DebugLoc in function test_max_be --  %tmp13.int = add nuw nsw i32 %tmp11.int, 1
WARNING: Instruction with empty DebugLoc in function test_max_be --  %tmp14 = icmp ult i32 %tmp13.int, 99999

WARNING: Instruction with empty DebugLoc in function test_max_be2 --  %indvar.conv = sitofp i32 %tmp11.int to float
WARNING: Instruction with empty DebugLoc in function test_max_be2 --  %tmp13.int = add nuw nsw i32 %tmp11.int, 1
WARNING: Instruction with empty DebugLoc in function test_max_be2 --  %tmp14 = icmp ult i32 %tmp13.int, 99999

WARNING: Instruction with empty DebugLoc in function test_max_be3 --  %indvar.conv = sitofp i32 %tmp11.int to float
WARNING: Instruction with empty DebugLoc in function test_max_be3 --  %tmp13.int = add nuw nsw i32 %tmp11.int, 1
WARNING: Instruction with empty DebugLoc in function test_max_be3 --  %tmp14 = icmp ult i32 %tmp13.int, 99999

WARNING: Instruction with empty DebugLoc in function pr55505_remove_redundant_fptosi_for_float_iv --  %indvar.conv = sitofp i32 %float.iv.int to float
WARNING: Instruction with empty DebugLoc in function pr55505_remove_redundant_fptosi_for_float_iv --  %float.iv.next.int = add nsw i32 %float.iv.int, -1
WARNING: Instruction with empty DebugLoc in function pr55505_remove_redundant_fptosi_for_float_iv --  %cmp = icmp ugt i32 %float.iv.next.int, 0
```

The missing debug location of the PHINode is not reported in the debugify's output.

I will patch these missing updates later.

### PR DESCRIPTION
**Pull request by Apochens**
fix #91436 . Adding debug location updates for the newly created `phi`, `add`, `icmp` and `sitofp` instructions, respectively.

### CODE REVIEW

**Comment by Apochens** [comment]
@SLTozer Sorry for this direct ping. Could you please help me check this PR and assign reviewers?

File: llvm/test/Transforms/IndVarSimplify/indvars-preversing-debugloc.ll
Comment (OCHyams) [inline llvm/test/Transforms/IndVarSimplify/indvars-preversing-debugloc.ll:27]: nit: could remove the dbg.values (and the declaration) as they just add noise

**Review by OCHyams — COMMENTED** [review]
Other than a few comments for the test this LGTM, thanks for the patch

File: llvm/test/Transforms/IndVarSimplify/indvars-preversing-debugloc.ll
Comment (jryans) [inline llvm/test/Transforms/IndVarSimplify/indvars-preversing-debugloc.ll:1]: Nit: Typo in file name: "preversing", should be "preserving"

File: llvm/test/Transforms/IndVarSimplify/indvars-preversing-debugloc.ll
Comment (OCHyams) [inline llvm/test/Transforms/IndVarSimplify/indvars-preversing-debugloc.ll:14]: Oops I never submitted my other comment.

IMO we should check the debug locations too, i.e., capture the !dbg number and check the contents. Something like this for each inst:

```
; CHECK: [[IV_INT:%.*]] = phi i32 [ 0, [[ENTRY:%.*]] ], [ [[DOTINT:%.*]], [[BB]] ], !dbg ![[dbg:[0-9+]]
...

; CHECK ![[dbg]] = !DILocation(line: 2,
```

**Review by jryans — COMMENTED** [review]
Thanks for working on this! 😄 

Change looks correct to me overall. The pass is creating replacement instructions, so it makes sense to copy over the original location as you have done here.

Just a small nit in the test to fix up.

File: llvm/test/Transforms/IndVarSimplify/indvars-preversing-debugloc.ll
Comment (Apochens) [reply llvm/test/Transforms/IndVarSimplify/indvars-preversing-debugloc.ll]: Ok, the checks are refined!

File: llvm/test/Transforms/IndVarSimplify/indvars-preversing-debugloc.ll
Comment (Apochens) [reply llvm/test/Transforms/IndVarSimplify/indvars-preversing-debugloc.ll]: Ok, I corrected the file name 😂

File: llvm/test/Transforms/IndVarSimplify/indvars-preserving-debugloc.ll
Comment (jryans) [inline llvm/test/Transforms/IndVarSimplify/indvars-preserving-debugloc.ll:1]: Apologies for another file name nit... Just noticed the file name starts with "indvars-", but it's already in the `IndVarSimplify` directory, so the "indvars-" prefix feels redundant (and most existing tests in that directory do not have it), so perhaps best to remove that here.

Comment (Apochens) [reply llvm/test/Transforms/IndVarSimplify/indvars-preserving-debugloc.ll]: Thanks for the suggestion!

**Review by jryans — APPROVED** [review]
Thanks for working on this! It now looks ready to go to me. 😄 

Since you don't have merge access, I'll plan to merge this after waiting a day for others to review if they wish to.

**Pull request merged**
